Fix dangling phi bug from loop-unroll

jaebaek opened the pull request:
When unrolling the following loop:
```
%const0 = OpConstant ...
%const1 = OpConstant ...
...
%LoopHeader = OpLabel
%phi0 = OpPhi %float %const0 %PreHeader %phi1 %Latch
%phi1 = OpPhi %float %const1 %PreHeader %x    %Latch
...
%LoopBody = OpLabel
%x = OpFSub %float %phi1 %phi0
...
```

the loop-unroll pass sets the value of `%phi0` as `%phi1` for the second
copy of the loop body. For example, the second copy of
`%x = OpFSub %float %phi1 %phi0` will be
`%y = OpFSub %float %x %phi1`.

Since all phi instructions for inductions will be removed after the
loop unrolling, `%phi1` will be a dead dangling phi. This CL keeps the
phi value from the previous loop iteration to map instruction operands
to the phi value instead of using phi.

For example, the second copy of `%x = OpFSub %float %phi1 %phi0` will be
`%y = OpFSub %float %x %const1` because the value of `%phi1` from the
first loop iteration is `%const1`.

jaebaek commented:
This will fix https://github.com/microsoft/DirectXShaderCompiler/issues/3685

dnovillo commented on source/opt/loop_unroller.cpp:
```diff
@@ -99,13 +100,18 @@ struct LoopUnrollState {
     previous_latch_block_ = new_latch_block;
     previous_condition_block_ = new_condition_block;
+    for (size_t index = 0; index < previous_phis_.size(); ++index) {
+      previous_phi_value_ids_[index] =
```
I suggest using a `map` here.

jaebaek replied:
Do you mean using a `map` for `previous_phi_value_ids_`?
`previous_phi_value_ids_` specify the value of phi for the previous loop iteration. It must use the same index as `previous_phis_`. Since `previous_phis_` is a vector, we should use a vector for it. Using `map` will make the code complicated.

jaebaek commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -3401,6 +3401,213 @@ OpFunctionEnd
+; CHECK: [[phi0_init:%\w+]] = OpUndef %float
+; CHECK: [[phi1_init:%\w+]] = OpFSub %float [[phi0_init]] [[phi0_init]]
+
+        %117 = OpFSub %float %158 %158
```
Updated.

jaebaek commented on test/opt/loop_optimizations/unroll_simple.cpp:
```diff
@@ -3401,6 +3401,213 @@ OpFunctionEnd
+; CHECK:      [[next_phi1_2:%\w+]] = OpFSub %float [[next_phi1_1]] [[next_phi1_0]]
+; CHECK-NEXT:                        OpFSub %float [[next_phi1_2]] [[next_phi1_1]]
+
+        %164 = OpPhi %float %float_1 %59 %130 %148
```
Done

jaebaek reviewed (commented):
I simplified the unit test examples.

s-perron reviewed (changes requested):
I'm not convinced this is the correct change.  I've looked at the code in  CopyBody, and that looks correct.  You should remember that this code is not always fully unrolling the loop, I believe this code will be incorrect when doing partial unrolling.  Below, I have attached the results of partially unrolling the loop using the orignal code ([old.log](https://github.com/KhronosGroup/SPIRV-Tools/files/6351579/old.log)) and the code in the pr ([new.log](https://github.com/KhronosGroup/SPIRV-Tools/files/6351582/new.log)).

Notice that the second OpFSub should use the result of the second OpPhi (%18) for it second parameter.  It does in old.log, but not in new.log.

jaebaek commented:
I removed my previous change (the code change in the initial commit) and added some code in `LoopUnrollerUtilsImpl::CloseUnrolledLoop` as we discussed offline.

Mainly, the code replaces the dangling phis with their initial values when closing the unrolled loop.
The partial unrolling concept is unclear to me. I am not sure this will generate the partial unrolling correctly.
